### DISCUSSION TITLE
Can I make a process take exclusive control of the CPU?

### DISCUSSION THREAD

**Discussion by rongcuid** [nextflow-io/nextflow#4715] [Q&A] [answered]
I would like certain processes to run exclusively. That is, when this process runs, no other process can execute in parallel. This is required as I am performing benchmarks. Is this possible in nextflow?

I have looked at the batch pattern and `maxForks`, but they seem to only limit the process itself, not other processes.

Example:

```
process exclusiveProcess {
  maxForks 1
  // maxSomething 1?
  "sleep 10"
}

process unrelatedProcess {
  "sleep 1"
}

workflow {
  exclusiveProcess | view
  unrelatedProcess | view
}
```
In the above example, the two processes are still executed in parallel. I want that `exclusiveProcess` actually runs exclusively.

**Comment by bentsherman** [comment]
You could set `executor.cpus` in your config file to limit how many CPUs are available to the pipeline, then set the `cpus` directive for the exclusive process such that it uses all the available CPUs and prevents any other task from running at the same time.

**Reply by rongcuid** [reply]
Only my exclusive process need to have exclusive access, and I would actually like other tasks to execute in parallel. I'm still new to nextflow, so I will try if your solution works

**Reply by rongcuid** [reply]
When I use the `cpus N` directive, is it possible to reference the maximum number here?

**Reply by bentsherman** [reply]
This is a bit of a hack but you could set `cpus` to `Runtime.availableProcessors()-1`

The -1 is because Nextflow reserves 1 cpu for itself

**Reply by rongcuid** [reply]
I get 
```
ERROR ~ Invalid method invocation `availableProcessors` with arguments:  on Runtime type`.
```

**Reply by rongcuid** [reply]
Also, wouldn't it be a problem if there's only a single processor in system?

I kind of fixed the above solution with:

```groovy
cpus {Runtime.getRuntime().availableProcessors()-1}
```

But I'll like to find a less hacky solution

**Reply by bentsherman** [reply]
I think what you'll have to do then is use a param:
```groovy
cpus (params.max_cpus - 1)
```

```bash
nextflow run main.nf --max_cpus $(nproc)
```

`nproc` is just a command that tells you how many CPUs you have but there are others

**Reply by bentsherman** [reply]
> Also, wouldn't it be a problem if there's only a single processor in system?

True, you'll have to do something like:
```groovy
cpus Math.max(1, params.max_cpus-1)
```

**Reply by rongcuid** [reply]
I just realize that I can just ask for all available processors instead of `max-1`, and it will not deadlock or anything. Is that the better solution?

**Reply by bentsherman** [reply]
How did you do it

**Reply by rongcuid** [reply]
```groovy
    maxForks 1
    cpus {Runtime.getRuntime().availableProcessors()}
```

**Reply by bentsherman** [reply]
I think that's fine. You probably still need to do `Math.max(1, Runtime.getRuntime().availableProcessors() - 1)` or else Nextflow might fail. And you shouldn't need the maxForks or curly braces

**Reply by rongcuid** [reply, marked as answer]
Here's my final solution:

```groovy
process exclusiveProcess {
  cpus { Runtime.runtime.availableProcessors() }
  "sleep 10"
}

process unrelatedProcess {
  "sleep 1"
}

workflow {
  exclusiveProcess | view
  unrelatedProcess | view
}

```
